File: backend/src/datatypes/study.py
from pydantic import BaseModel
from .text import Explanation
from enum import Enum
from typing import Optional

class ExplanationMethod(Enum):
    PSHAP = "PSHAP"
    MEXGEN = "MEXGEN"
    HUMAN = "HUMAN"

class PersonalInformation(BaseModel):
    userName: str
    educationalBackground: Optional[str] = None
    occupation: Optional[str] = None
    researchField: Optional[str] = None

class PersonalInformationPostRequest(BaseModel):
    personalInformation: PersonalInformation
    confirmSamePerson: Optional[bool] = False
    confirmChangeInformation: Optional[bool] = False

class PersonalInformationPostResponse(BaseModel):
    userAlreadyRegistered: bool
    nameExists: bool
    nameChange: bool


class ComparableExplanation(BaseModel):
    explanation: Explanation
    userId: str
    taskId: str
    explanationMethod: ExplanationMethod

class ExplanationComparisonArrayResponse(BaseModel):
    explanationComparisonArray: list[list[ComparableExplanation]]

class ComparisonResult(BaseModel):
    preferredMethod: ExplanationMethod
    userId : str
    taskId: str

class ComparisonStudyResults(BaseModel):
    comparisonResults: list[ComparisonResult]

File: frontend/src/datatypes/study.ts
import { Explanation } from "./network";

export enum ExplanationMethod {
    PSHAP = "PSHAP",
    MEXGEN = "MEXGEN",
    HUMAN = "HUMAN",
}

export interface PersonalInformation {
    userName: string;
    educationalBackground?: string;
    occupation?: string;
    researchField?: string;
    // TODO: What other information should be collected?
}
export interface PersonalInformationPostRequest {
    personalInformation: PersonalInformation;
    confirmSamePerson: boolean; // If get same user name, need to confirm whether it is the same person on a new device or a different person.
    confirmChangeInformation: boolean; // If cookie already has information registered need to check whether person really wants to update the information.

}
export interface PersonalInformationPostResponse {
    userAlreadyRegistered: boolean;
    nameExists: boolean;
    nameChange: boolean;
}

export interface ComparableExplanation {
    explanation: Explanation;
    userId: string;
    explanationMethod: ExplanationMethod;
}

export interface ExplanationComparisonArrayResponse {
    explanationComparisonArray: ComparableExplanation[][];
}


export interface ComparisonResult {
    preferredMethod: ExplanationMethod;
    userId: string;
    taskId: string;
}
export interface ComparisonStudyResults {
    comparisonResults: ComparisonResult[];
}

export interface CompletedTasks {
    currentLevel: number;
    completedTask_ids: string[];
}

